File: client/src/utils/emailValidator.js
/**
 * Disposable/temporary email domain blocklist.
 * Only trusted email providers are allowed for sign-up.
 */

const ALLOWED_DOMAINS = new Set([
  // Google
  'gmail.com', 'googlemail.com',
  // Microsoft
  'outlook.com', 'hotmail.com', 'live.com', 'msn.com',
  // Yahoo
  'yahoo.com', 'yahoo.co.uk', 'yahoo.co.in', 'ymail.com', 'rocketmail.com',
  // Apple
  'icloud.com', 'me.com', 'mac.com',
  // ProtonMail
  'protonmail.com', 'proton.me', 'pm.me',
  // Zoho
  'zoho.com', 'zohomail.com',
  // AOL
  'aol.com',
  // Tutanota
  'tutanota.com', 'tutamail.com', 'tuta.io',
  // FastMail
  'fastmail.com', 'fastmail.fm',
  // GMX
  'gmx.com', 'gmx.net',
  // Mail.com
  'mail.com', 'email.com',
  // Educational (generic patterns — accept any .edu domain)
  // Handled separately below
]);

/**
 * Check if an email uses a trusted/allowed domain.
 * Returns { valid: boolean, reason?: string }
 */
export function validateEmailDomain(email) {
  if (!email || typeof email !== 'string') {
    return { valid: false, reason: 'Email is required.' };
  }

  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!emailRegex.test(email)) {
    return { valid: false, reason: 'Invalid email format.' };
  }

  const domain = email.split('@')[1].toLowerCase();

  // Allow educational domains (.edu, .ac.*)
  if (domain.endsWith('.edu') || domain.includes('.ac.')) {
    return { valid: true };
  }

  // Allow corporate/organization domains (.org, .gov)
  if (domain.endsWith('.org') || domain.endsWith('.gov')) {
    return { valid: true };
  }

  if (ALLOWED_DOMAINS.has(domain)) {
    return { valid: true };
  }

  return {
    valid: false,
    reason: 'Disposable or temporary email addresses are not allowed. Please use a trusted email provider (Gmail, Outlook, Yahoo, etc.).',
  };
}

File: server/utils/email_validator.py
"""
Server-side disposable email validation.
Mirrors the client-side logic but acts as the authoritative check.
"""

ALLOWED_DOMAINS = {
    # Google
    "gmail.com", "googlemail.com",
    # Microsoft
    "outlook.com", "hotmail.com", "live.com", "msn.com",
    # Yahoo
    "yahoo.com", "yahoo.co.uk", "yahoo.co.in", "ymail.com", "rocketmail.com",
    # Apple
    "icloud.com", "me.com", "mac.com",
    # ProtonMail
    "protonmail.com", "proton.me", "pm.me",
    # Zoho
    "zoho.com", "zohomail.com",
    # AOL
    "aol.com",
    # Tutanota
    "tutanota.com", "tutamail.com", "tuta.io",
    # FastMail
    "fastmail.com", "fastmail.fm",
    # GMX
    "gmx.com", "gmx.net",
    # Mail.com
    "mail.com", "email.com",
}


def validate_email_domain(email: str) -> tuple[bool, str]:
    """
    Check if an email uses a trusted domain.
    Returns (is_valid: bool, reason: str).
    """
    if not email or "@" not in email:
        return False, "Invalid email address."

    domain = email.split("@")[1].lower().strip()

    # Allow educational domains
    if domain.endswith(".edu") or ".ac." in domain:
        return True, ""

    # Allow .org and .gov
    if domain.endswith(".org") or domain.endswith(".gov"):
        return True, ""

    if domain in ALLOWED_DOMAINS:
        return True, ""

    return False, (
        "Disposable or temporary email addresses are not allowed. "
        "Please use a trusted email provider (Gmail, Outlook, Yahoo, etc.)."
    )
